tests/nn/decoding/expected_risk_minimization_test.py

```diff
@@ -0,0 +1,103 @@
+                                                               state.action_history,
+                                                               state.score,
+                                                               allowed_actions):
+
+            for action in actions:
+                next_item = action_history[-1] + action
+                new_history = action_history + [next_item]
+                # For every action taken, we reduce the score by 1.
+                new_score = score - 1
+                new_state = SimpleDecoderState([batch_index],
+                                               [new_history],
+                                               [new_score])
+                indexed_next_states[batch_index].append(new_state)
+        next_states: List[SimpleDecoderState] = []
+        for batch_next_states in indexed_next_states.values():
+            if max_actions is not None:
+                batch_next_states = batch_next_states[:max_actions]
+            next_states.extend(batch_next_states)
+        return next_states
+
+
+class TestExpectedRiskMinimization(AllenNlpTestCase):
+    def setUp(self):
+        super(TestExpectedRiskMinimization, self).setUp()
+        self.initial_state = SimpleDecoderState([0], [[0]], [Variable(torch.Tensor([0.0]))])
+        self.decoder_step = SimpleDecoderStep()
+        # Cost is the number of odd elements in the action history.
+        self.supervision = lambda state: Variable(torch.Tensor([sum([x%2 != 0 for x in
+                                                                     state.action_history[0]])]))
+        # High beam size ensures exhaustive search.
+        self.trainer = ExpectedRiskMinimization(beam_size=100,
+                                                normalize_by_length=False,
+                                                max_decoding_steps=10)
+
+    def test_get_finished_states(self):
+        finished_states = self.trainer._get_finished_states(self.initial_state, self.decoder_step)
+        state_info = []
+        for state in finished_states:
+            state_info.append((state.action_history[0], int(state.score[0].data)))
+        # There will be exactly five finished states with the following paths. Each score is the
+        # negative of one less than the number of elements in the action history.
+        assert len(finished_states) == 5
+        assert ([0, 2, 4], -2) in state_info
+        assert ([0, 1, 2, 4], -3) in state_info
+        assert ([0, 1, 3, 4], -3) in state_info
+        assert ([0, 2, 3, 4], -3) in state_info
+        assert ([0, 1, 2, 3, 4], -4) in state_info
+
+    def test_decode(self):
+        decoded_info = self.trainer.decode(self.initial_state, self.decoder_step, self.supervision)
+        # The best state corresponds to the shortest path.
+        assert decoded_info['best_action_sequence'][0] == [0, 2, 4]
+        assert_almost_equal(decoded_info['loss'].data.numpy(), 0.7781, decimal=4)
```

Comment:
Where does this loss come from?  Oh, yeah, it's the average of the stuff on the beam.  Some note here would be good.